packages/internal/src/utils.js

```diff
@@ -48,16 +42,16 @@ const { details: X } = assert;
  * if all the mapped values are Passable, then the returned object will be
  * a CopyRecord.
  *
- * @template T
- * @template U
- * @param {Record<string,T>} original
- * @param {(value: T, key?: string) => U} mapFn
- * @returns {Record<string,U>}
+ * @template {Record<string, any>} O
+ * @param {O} original
+ * @template R map result
+ * @param {(value: O[keyof O], key?: keyof O) => R} mapFn
```

Comment:
> Is the ? right? It seems right at first, but it is in the backwards direction.

You're right, this was an error. It should be `key` as the function passed in will always be called with it. I've fixed and added tests

Reply:
> We are not expression as much precision as we'd like, perhaps because more precision is impossible in TS. But the imprecision is that `R` types the values of all the properties of the result.

Correct. This is treating the object as a Record of keys to some single type, which will be a union of all the types that are present. So the mapFn has to take all those. What it returns is called R, which could also be a union.

> say that the output type of the value associated with each key depends on the type of the input value associated with that same key?

I tried that but the `mapFn` is what defines the actual mapping and it would have to define the input/output for key key. I don't think that's practical.